### ISSUE TITLE
Comprehensive overhaul of `factorint`

### ISSUE THREAD

**Issue by haru-44** [sympy/sympy#27509]
## Introduction
`sympy.ntheory.factor_.factorint` is a function that performs integer factorization. Although there are several issues with its implementation, fundamental changes have been avoided due to its complexity. In this document, I aim to organize the existing problems and explore potential improvements in its implementation.

#### Recursive execution
Every time a factor is found, `factorint` is recursively called. This is a natural implementation since non-prime factors need to be further decomposed. However, since trial division is performed repeatedly, unnecessary computations occur.

#### Inadequate parameters for Pollard's rho algorithm
When executing `pollard_rho`, specific parameters are used. However, when called from `factorint`, these parameters are fixed, leading to redundant computations.

#### Selection of algorithms
When using multiple factorization algorithms, the choice of parameters has not been sufficiently considered. This has led to issues such as the one reported in #27424 .

### Potential for incorrect answers
The `use_*` options allow users to specify which factorization algorithms to use. However, the implementation does not properly handle cases where certain algorithms are disabled. For example, if only ECM is enabled, incorrect results may be returned.

### Options
When `verbose=True`, log output is printed using `print`, but using a logger would be a better approach. In the first place, it is questionable whether a logging feature is necessary.

An option to yield factors as they are found would be useful in cases like checking whether a number is square-free (see #27271).

Although the `limit` option has its use cases, it unnecessarily complicates the implementation.

### Potential for speedup
`flint.factor` is a factorization function provided by `python-flint`. Generally, `flint` functions are faster, so if `flint` is installed, it should be used. Similar external functions in SymPy are grouped under `sympy.external.gmpy`, and following this approach would maintain consistency. However, this would require moving many factorization-related functions to `sympy.external`, which is undesirable since they are used by other `ntheory` functions. Therefore, if `flint.factor` is to be incorporated, it should be called within the `factorint` function.

### Initial trial division
Initially, trial division using very small primes is performed. This should be executed even if `use_trial` is set to `False`, ensuring that algorithms relying on odd-numbered inputs can operate correctly. The threshold for this needs to be determined.

### Trial division for small `n`
When `n` is small (e.g., `n < 2**32`), simple trial division is often faster than running more sophisticated factorization algorithms. The exact threshold needs further evaluation.

### Managing unfactored numbers
Numbers that are not yet fully factored will be managed using a structure like `unfactored: list[tuple[int, int]]`. This tracks expressions of the form $m_1^{e_1} m_2^{e_2} \cdots m_k^{e_k}$, where each $m_i$ is neither prime nor a perfect power. This approach avoids recursive calls and redundant computations.

#### T-level
A metric known as the t-level is used to measure the progress of ECM. If a factor of size $2^t$ exists, it should be found with a probability of $1 - e^{-1}$. This metric is based on empirical evidence rather than theoretical foundations. Similarly, the t-level can be known by experimentation with the p-1 and rho methods. Using t-level, the progress of different algorithms can be evaluated using the same indicator.

For reference: [OEIS - T-Levels](https://oeis.org/wiki/OEIS_sequences_needing_factors#T-levels)

#### ECM parameters
It is likely best to use empirically known parameters.

For reference: [Choosing the Best Parameters for ECM](https://www.rieselprime.de/ziki/Elliptic_curve_method#Choosing_the_best_parameters_for_ECM)

#### Prioritization
By comparing the t-levels of p-1, rho, and ECM, the next parameters to test can be determined. However, execution time must also be considered when choosing which algorithm to use. At this stage, I do not have a clear idea of the best approach for this.

## Conclusion
Many aspects of this sketch are speculative, and practical implementation may reveal that some ideas are unfeasible. Thus, revisions will likely be necessary. In fact, my thoughts on this topic are still quite unorganized. Given these circumstances, it will likely take some time before `factorint` can actually be overhauled.

**Comment by haru-44** [comment]
The rho method does not require computing the gcd every time. Additionally, unless it fails, we can resume from where we left off, so we should save `U` and `V` and restart from there next time.

```python
rho_param = {}

def rho(n):
    CONT_NUM = 128
    if n in rho_param:
        s, a, U, V, c = rho_param[n]
    else:
        s = 2
        a = 1
        U = V = s
        c = 0

    for _ in range(100):
        t = 1
        prev = U, V
        for _ in range(CONT_NUM):
            U = pow(U, 2, n) + a
            V = pow(pow(V, 2, n) + a, 2, n) + a
            t = t*(U - V) % n
        c += 1
        g = math.gcd(t, n)
        if g == 1:
            continue
        if g != n:
            return g
        U, V = prev
        for _ in range(CONT_NUM):
            U = pow(U, 2, n) + a
            V = pow(pow(V, 2, n) + a, 2, n) + a
            g = math.gcd(U - V, n)
            if g == 1:
                continue
            if g != n:
                return g
            break
        # If it fails, use another parameter
        s += 1
        a += 1
        U = V = s
        c = 0
    rho_param[n] = s, a, U, V, c
```

Additionally, by experimenting, we can estimate how quickly a prime factor can be discovered. If the loop is executed k times, we can expect to find a factor of approximately $2^{2\log_2{k} - 0.35}$ in size.

<details>

```python
import math
import random
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sympy import sieve, E, nextprime

sieve.extend(1_000_000)
results = []
q = nextprime(10**100)

for _ in range(100_000):
    r = random.uniform(10, 30)
    p = nextprime(2**r)
    r = math.log2(p)
    n = p * q
    s = random.randrange(n)
    a = random.randrange(1, n - 2)
    V = U = s
    k = 0
    while True:
        U = pow(U, 2, n) + a
        V = pow(pow(V, 2, n) + a, 2, n) + a
        g = math.gcd(U - V, n)
        k += 1
        if 1 < g < n:
            results.append((r, k))
            break

df = pd.DataFrame(results)
df.columns = ['lg_p', 'k']
df['cut'] = pd.cut(df['lg_p'], 100)
df = df.groupby('cut', as_index=False, observed=False)['k'].quantile(1-(E**-1))
df['t'] = df['cut'].map(lambda x: x.mid)
df['lg_k'] = np.log2(df['k'])

model = LinearRegression()
model.fit(df[['lg_k']], df['t'])
print(model.coef_, model.intercept_)
# [1.99507758] -0.34703705868230017
```

</details>

**Comment by ayush-bothra** [comment]
hello,
I can start work on getting flint into the `factorint` function

**Comment by haru-44** [comment]
Anyone may carry out the implementation, but please do so with a clear understanding of the underlying issue.
Simply adding a call to `flint.factorint` inside `factorint` is not sufficient. Once `flint.factorint` is invoked, SymPy can no longer control the process until the factorization is complete, and it is unclear when that will happen. In contrast, `factorint` provides a `limit` option that allows the function to return without fully factoring the integer.
If you have ideas on how to address this, please put them on the table for discussion.

**Comment by oscarbenjamin** [comment]
The python-flint factor function does have a `trial_limit` parameter but it is not precisely equivalent to what sympy's factorint does with limit. It could also be possible to make changes in python-flint to give something more directly analogous or to adapt the output from python-flint:
```
In [2]: flint.fmpz(2*10*123456789).factor(trial_limit=5)
Out[2]: [(2, 2), (3, 2), (5, 1), (3607, 1), (3803, 1)]

In [3]: factorint(2*10*123456789, limit=5)
Out[3]: {2: 2, 3: 2, 5: 1, 13717421: 1}
```

**Comment by ayush-bothra** [comment]
safest option might be to add a check before running flint to see if these parameters were passed/are not None
maybe we shld keep those cases to sympy for now?
since haru_44 has also provided a possible refactor for the section (removing the repeated trial division if I am not mistaken)
that can be done as a separate work after adding flint functionality??

**Comment by asmeurer** [comment]
I find the verbose output to be useful. It shows that the function is doing something when it runs for a long time. It also lets you see the existing factors as they are found (although an option to return a generator could also solve this problem). I'm not sure I agree that it needs to be using a real logger. The feature exists primarily for interactive human feedback.

**Comment by haru-44** [comment]
> The python-flint factor function does have a `trial_limit` parameter but it is not precisely equivalent to what sympy's factorint does with limit.

In both SymPy and flint, the meaning of the value passed to `limit` or `trial_limit` is ambiguous. It does not mean that only prime factors up to that value will appear in the result. If the point is simply that there is a mechanism to stop factorization partway through, then using `trial_limit` might be acceptable.

> safest option might be to add a check before running flint to see if these parameters were passed/are not None
> maybe we shld keep those cases to sympy for now?

We should also think about where in `factorint` the call to `factor` should happen. Should the initial trial division and primality testing be done on the SymPy side first? For simple cases, solving them using SymPy alone might be faster since it avoids overhead, but it is unclear how much benefit this provides overall.

Incremental improvements are not necessarily bad, but I am concerned that features and options added one by one may make future refactoring more difficult.

> I find the verbose output to be useful. It shows that the function is doing something when it runs for a long time. It also lets you see the existing factors as they are found (although an option to return a generator could also solve this problem). I'm not sure I agree that it needs to be using a real logger. The feature exists primarily for interactive human feedback.

From my perspective, I would like to consider the options in the order: generator, logging, and then print. For displaying progress, logging is more appropriate than print, and for interactive feedback, a generator is more suitable than print.

**Comment by ayush-bothra** [comment]
> We should also think about where in `factorint` the call to `factor` should happen. Should the initial trial division and primality testing be done on the SymPy side first? For simple cases, solving them using SymPy alone might be faster since it avoids overhead, but it is unclear how much benefit this provides overall.
> 
> Incremental improvements are not necessarily bad, but I am concerned that features and options added one by one may make future refactoring more difficult.

I think we can add a helper function whose sole responsibility is to call flint, convert the value into a flint object and return the computed ans as a dict, it will know nothing else
Inside the `_factorint` function, after all the checks for other types like Mul etc, and after its done with the trial division
we can add a check as follows:
```text
if limit is None && n (the num to work on) > 1 && not verbose:
    check if flint exists
        call defined helper to solve using flint
    return the dict+previously found factors directly
else: pollards rho and ECM can be run here
```
This separates flint from the function and we only have to change things inside that function OR within `_factorint` conditions
we run the trial divisions before calling the helper, allowing simpler or smoother numbers to get factored without object overhead
finally we are able to use flint for simple queries which only require the factors and care about nothing else

**Comment by oscarbenjamin** [comment]
One thing to be careful of is that `factorint` is used for this kind of thing:
```python
In [1]: sqrt(12)
Out[1]: 2⋅√3
```
The `limit` controls which integers are factored:
https://github.com/sympy/sympy/blob/4f3e824d490b62164e347db15e6e61e311f0d156/sympy/core/numbers.py#L2132
Hence:
```python
In [1]: p1, p2 = 3, 5

In [2]: sqrt(p1*p2**2) - sqrt(p1)*p2
Out[2]: 0

In [3]: p1 = nextprime(2**15)

In [4]: p2 = nextprime(p1)

In [5]: sqrt(p1*p2**2) - sqrt(p1)*p2
Out[5]: -32779⋅√32771 + √35211221762411
```
It would not be good for these things to behave differently depending on whether python-flint is installed or is being used for factorisation.

**Comment by haru-44** [comment]
> It would not be good for these things to behave differently depending on whether python-flint is installed or is being used for factorisation.

You're absolutely right. However, making the behavior consistent while using flint in an integrated way is starting to seem difficult.

**Comment by oscarbenjamin** [comment]
I think it would be worth separating out a more limited version of `factorint` for use in `Integer._eval_power` and making it as fast as possible including by using python-flint when possible.

It just needs to have a clearly defined behaviour like "finds all prime factors less than `2^15` and returns the remaining factor as a single integer". Then I think it is possible at least to wrap python-flint for that behaviour:
```python
from flint import fmpz

def factorint_flint(n: int) -> dict[int, int]:
    """Find prime factors less than 2**15"""
    factors_fmpz = fmpz(n).factor(trial_limit=2**15)
    remainder = fmpz(1)
    factors: dict[int, int] = {}
    for p, e in factors_fmpz:
        if p < 2**15:
            factors[int(p)] = int(e)
        else:
            remainder *= p**e
    if remainder != 1:
        factors[int(remainder)] = 1
    return factors
```
Then
```python
In [3]: factorint_flint(1234567890)
Out[3]: {2: 1, 3: 2, 5: 1, 3607: 1, 3803: 1}

In [4]: Integer(1234567890).factors(limit=2**15)
Out[4]: {2: 1, 3: 2, 5: 1, 3607: 1, 3803: 1}
```

**Comment by oscarbenjamin** [comment]
Actually
```python
from flint import fmpz

def factorint_flint(n: int) -> dict[int, int]:
    """Find prime factors less than 2**15"""
    if n == 0:
        return {0: 1}
    factors: dict[int, int] = {}
    factors_fmpz = fmpz(n).factor(trial_limit=2**15)
    remainder = fmpz(1)
    for p, e in factors_fmpz:
        if p < 2**15:
            factors[int(p)] = int(e)
        else:
            remainder *= p**e
    if remainder != 1:
        factors[int(remainder)] = 1
    if n < 0:
        factors[-1] = 1
    return factors
```
Turns out that python-flint does not add the `-1` for negative numbers and crashes on zero. This version produces the exact same output with the exact same dict order:
```
In [18]: for n in range(-2**20, 2**20):
    ...:     assert str(Integer(n).factors(limit=2**15)) == str(factorint_flint(n))
    ...:
```

**Comment by haru-44** [comment]
I don't believe `Integer._eval_power` really needs advanced factorization algorithms. A function that only does trial division might already be adequate.

In addition, it is unclear to me how much performance gain we would get from using flint here. While I am not against applying flint to small integers, the advantage should be more significant for larger inputs.

**Comment by oscarbenjamin** [comment]
The `factorint_flint` function I showed above is faster than `factorint` but not by as much as I expected at least for the numbers I tried. It seems about 4x faster for small ints and about 2x faster for larger ones.

I imagine that at the very least a more basic version of factorint for small factors could be faster just by skipping out all the random other stuff that `factorint` et al do:

- `Integer._eval_power` calls `Rational.factors`
- `Rational.factors` imports and calls `factorrat` passing a bunch of arguments that aren't needed and then copys the returned dict for no apparent reason.
- `factorrat` calls `factorint` twice, converts the result to a defaultdict and then back to a dict.
- `factorint` checks loads of stuff before it gets to `_factorint_small` and then does a load of other stuff as well.

Skipping all of that and going straight to `_factorint_small` would already save a bunch of time.

**Comment by ayush-bothra** [comment]
```python
    if limit is None and n > 1 and not verbose and visual is None:
        fact = _flint_factorint(n)
        if fact is not None:
            for key, value in fact.items():
                factors[key] = factors.get(key, 0) + value
            factor_cache[n] = min(factors.keys())
            return factors
```

This is the conditions I used after factorint has made checks for other types and negative and zero n

```python
def test_flint_factorization():
    flint = import_module('flint')
    if not flint:
        skip("python-flint not installed")
    p = nextprime(10**30)
    q = nextprime(10**33)
    n = p * q
    result = factorint(n)
    assert result == {p : 1, q : 1}
    n_mixed = 2 * p * q
    assert factorint(n_mixed) == {2: 1, p: 1, q: 1}
    
    small_n = 123456789123456789
    expected = {
        3: 2, 
        7: 1, 
        11: 1, 
        13: 1, 
        19: 1, 
        3607: 1, 
        3803: 1, 
        52579: 1
    }
    
    assert factorint(small_n, verbose=True) == expected
    
    assert factorint(0) == {0 : 1}
    assert factorint(-10) == {-1: 1, 2: 1, 5: 1}
    
    p1 = nextprime(2**15)
    p2 = nextprime(p1)
    expr = sqrt(p1 * p2**2) - sqrt(p1) * p2
    assert expr != 0 
```


these are a few tests that I ran 
firs one is purely to see if flint is able to solve it
second one is to check if we have a case where trial division reduces the number, is my code able to combine everything together
next one was to check if code falls back to sympy if any of the parameters are set
next are a few checks to see if simplification happened and checks for negatives

```text
rootdir: /home/yush/sympy
configfile: pyproject.toml
plugins: hypothesis-6.150.0
collected 35 items / 1 deselected / 34 selected                                                                                                                   

sympy/ntheory/tests/test_factor_.py ..................................                                                                                      [100%]

================================================================ 34 passed, 1 deselected in 8.69s =================================================================
```
and for the verbosity
```text
======================================================================= test process starts =======================================================================
executable:         /home/yush/sympy/.venv/bin/python3  (3.12.3-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       flint 
numpy:              None
random seed:        68614788
hash randomization: on (PYTHONHASHSEED=4034989907)

sympy/ntheory/tests/test_factor_.py[1] Factoring 123456789123456789
Trial division with ints [2 ... 32768] and fail_max=600
        3 ** 2
        7 ** 1
        11 ** 1
        13 ** 1
        19 ** 1
Check for termination
Trial division with primes [1823 ... 3646]
        3607 ** 1
Check for termination
Trial division with primes [3646 ... 7292]
        3803 ** 1
Check for termination
Factorization is complete.
.                                                                                                                       [OK]

__________________________________________________________________________ slowest tests __________________________________________________________________________
sympy/ntheory/tests/test_factor_.py::test_flint_factorization - Took 12.042 seconds
============================================================ tests finished: 1 passed, in 9.48 seconds ============================================================
```

I havent put a PR simply because if changes are to be made to some other sections and if it affects this code then it will have to be changed anyways
are there any other checks or changes I shld make and test?

**Comment by haru-44** [comment]
flint implements individual factorisation algorithms such as `fmpz_factor_trial_range`. If python-flint provides wrappers for these, it would be ideal to use them.

As an initial fix, I think it would be acceptable to create a function that replaces `_factorint_small`.

```python
def _factorint_small_new(n: int, limit: int, strict: bool) -> dict[int, int]:
    if _SYMPY_GROUND_TYPES == 'flint':
        factors_fmpz = fmpz(n).factor(trial_limit=limit)
        # convert to a dict
        if not strict:
            return factors_fmpz
        # extract only prime factors <= limit
        return factors_fmpz
    # If flint is not used, perform trial division as in the existing `_factorint_small`
```

The `strict` option is introduced to account for the possibility that prime factors larger than `limit` may be found.
